### PR TITLE
Fix white space handling while parsing LRC with word time tags

### LINKED ISSUE #83: Removing time tag leaves whitespace behind
Using version `2025.228.1`

With a line such as 

```
[00:06.84] <00:06.84> Every <00:07.20>   <00:07.56> night <00:07.87>   <00:08.19> that <00:08.46>   <00:08.79> goes <00:09.19>   <00:09.59> between
```

LrcParser.Decode() outputs text as

```
Every     night     that     goes     between
```
instead of
```
Every night that goes between
```

Full .elrc contents for testing

```
[00:06.84] <00:06.84> Every <00:07.20>   <00:07.56> night <00:07.87>   <00:08.19> that <00:08.46>   <00:08.79> goes <00:09.19>   <00:09.59> between
[00:14.69] <00:14.69> I <00:14.78>   <00:14.87> feel <00:15.15>   <00:15.44> a <00:15.54>   <00:15.65> little <00:15.96>   <00:16.28> less
[00:20.98] <00:20.98> As <00:21.10>   <00:21.22> you <00:21.27>   <00:21.31> slowly <00:21.79>   <00:22.27> go <00:22.57>   <00:22.87> away <00:23.74>   <00:24.19> from <00:24.51>   <00:24.82> me
[00:28.73] <00:28.73> This <00:28.91>   <00:29.09> is <00:29.22>   <00:29.36> only <00:29.75>   <00:30.14> another <00:30.92>   <00:31.25> test
[00:36.11] <00:36.11> Every <00:36.44>   <00:36.77> night <00:37.14>   <00:37.52> you <00:37.73>   <00:38.00> do <00:38.24>   <00:38.48> not <00:38.93>   <00:39.41> come
[00:43.56] <00:43.56> Your <00:43.65>   <00:43.74> softness <00:44.21>   <00:44.69> fades <00:45.01>   <00:45.33> away
[00:50.29] <00:50.29> Did <00:50.42>   <00:50.56> I <00:50.70>   <00:50.85> ever <00:51.41>   <00:51.97> really <00:52.48>   <00:52.99> care <00:53.56>   <00:53.86> that <00:54.09>   <00:54.34> much
[00:58.07] <00:58.07> Is <00:58.20>   <00:58.34> there <00:58.48>   <00:58.63> anything <00:59.44>   <00:59.69> left <00:59.94>   <01:00.20> to <01:00.30>   <01:00.41> say?
[01:05.59] <01:05.59> Every <01:06.58>   <01:07.78> hour <01:08.05>   <01:08.32> of <01:08.92>   <01:09.97> fear <01:10.39>   <01:10.81> I <01:11.41>   <01:11.47> spend
[01:13.84] <01:13.84> My <01:13.99>   <01:14.14> body <01:14.57>   <01:15.01> tries <01:15.32>   <01:15.64> to <01:15.71>   <01:15.79> cry
[01:18.60] <01:18.60> Living <01:19.41>   <01:20.79> through <01:21.12>   <01:21.45> each <01:21.90>   <01:23.13> empty <01:23.83>   <01:24.54> night
[01:27.51] <01:27.51> A <01:27.60>   <01:27.69> deadly <01:28.12>   <01:28.56> call <01:28.90>   <01:29.25> inside
[01:34.29] <01:34.29> I <01:34.39>   <01:34.50> haven't <01:35.37>   <01:35.64> felt <01:35.92>   <01:36.21> this <01:36.34>   <01:36.48> way <01:36.79>   <01:37.11> I <01:37.35>   <01:37.62> feel
[01:42.11] <01:42.11> Since <01:42.27>   <01:42.44> many <01:42.71>   <01:42.98> a <01:43.08>   <01:43.19> years <01:43.46>   <01:43.85> ago
[01:48.92] <01:48.92> But <01:49.08>   <01:49.25> in <01:49.36>   <01:49.48> those <01:49.78>   <01:50.09> years <01:50.50>   <01:50.92> and <01:51.07>   <01:51.23> the <01:51.30>   <01:51.38> lifetime's <01:51.99>   <01:52.61> past
[01:56.60] <01:56.60> I <01:56.69>   <01:56.78> did <01:56.91>   <01:57.05> not <01:57.50>   <01:57.92> deal <01:58.22>   <01:58.52> with <01:58.68>   <01:58.85> the <01:58.91>   <01:58.97> road
[02:03.13] <02:03.13> And <02:03.25>   <02:03.37> I <02:03.45>   <02:03.55> did <02:03.68>   <02:03.81> not <02:04.20>   <02:04.60> deal <02:04.94>   <02:05.29> with <02:05.45>   <02:05.62> you, <02:05.90>   <02:06.19> I <02:06.50>   <02:06.82> know
[02:10.95] <02:10.95> Though <02:11.11>   <02:11.28> the <02:11.35>   <02:11.43> love <02:11.79>   <02:12.15> has <02:12.32>   <02:12.39> always <02:13.06>   <02:13.74> been
[02:17.70] <02:17.70> So <02:17.91>   <02:18.12> I <02:18.16>   <02:18.21> search <02:18.55>   <02:18.90> to <02:19.03>   <02:19.17> find <02:19.51>   <02:19.86> an <02:20.14>   <02:20.43> answer <02:21.24>   <02:21.71> there
[02:25.72] <02:25.72> So <02:25.96>   <02:26.20> I <02:26.39>   <02:26.59> can <02:27.04>   <02:27.19> truly <02:27.65>   <02:28.12> win
[02:33.02] <02:33.02> Every <02:34.04>   <02:35.29> hour <02:35.77>   <02:35.84> of <02:36.38>   <02:37.43> fear <02:37.92>   <02:38.42> I <02:38.69>   <02:38.96> spend
[02:41.35] <02:41.35> My <02:41.63>   <02:41.91> body <02:42.35>   <02:42.79> tries <02:43.09>   <02:43.39> to <02:43.48>   <02:43.57> cry
[02:46.32] <02:46.32> Living <02:47.04>   <02:48.54> through <02:48.88>   <02:49.23> each <02:49.56>   <02:50.85> empty <02:51.55>   <02:52.26> night
[02:55.06] <02:55.06> A <02:55.13>   <02:55.21> deadly <02:55.64>   <02:56.08> call <02:56.42>   <02:56.77> inside
[03:01.50] <03:01.50> So <03:01.74>   <03:01.98> I <03:02.04>   <03:02.10> try <03:02.42>   <03:02.76> to <03:02.81>   <03:02.87> say <03:03.45>   <03:04.20> goodbye, <03:04.53>   <03:04.86> my <03:05.02>   <03:05.19> friend
[03:09.09] <03:09.09> I'd <03:09.19>   <03:09.30> like <03:09.42>   <03:09.54> to <03:09.70>   <03:09.87> leave <03:10.02>   <03:10.17> you <03:10.29>   <03:10.41> with <03:10.71>   <03:10.89> something <03:11.32>   <03:11.76> warm
[03:16.25] <03:16.25> But <03:16.34>   <03:16.43> never <03:16.77>   <03:17.12> have <03:17.25>   <03:17.38> I <03:17.44>   <03:17.51> been <03:17.73>   <03:17.96> a <03:18.14>   <03:18.32> blue <03:18.62>   <03:18.92> calm <03:19.58>   <03:19.76> sea
[03:24.18] <03:24.18> I <03:24.31>   <03:24.45> have <03:24.57>   <03:24.69> always <03:25.59>   <03:28.02> been <03:28.19>   <03:28.38> a <03:28.51>   <03:28.65> storm
[03:33.69] <03:33.69> Always <03:35.43>   <03:36.99> been <03:37.18>   <03:37.38> a <03:37.50>   <03:37.62> storm
[03:41.45] <03:41.45> Ooh, <03:41.99>   <03:42.53> always <03:44.42>   <03:45.86> been <03:46.05>   <03:46.25> a <03:46.41>   <03:46.58> storm
[03:50.34] <03:50.34> I <03:50.49>   <03:50.64> have <03:51.07>   <03:51.51> always <03:53.31>   <03:54.78> been <03:54.96>   <03:55.14> a <03:55.47>   <03:55.80> storm
```

### PR DESCRIPTION
**Pull request by Maxr1998**
- Correctly handle white space between word time tags
- Support parsing time tags that surround each non-blank segment

Fixes #83.

### CODE REVIEW

File: LrcParser/Parser/Lrc/Lines/LrcLyricParser.cs [hunk 1]
```diff
@@ -9,8 +9,9 @@ namespace LrcParser.Parser.Lrc.Lines;
 
 public class LrcLyricParser : SingleLineParser<LrcLyric>
 {
-    public override bool CanDecode(string text)
-        => !string.IsNullOrWhiteSpace(text);
+    public override bool CanDecode(string line) =>
+        !string.IsNullOrWhiteSpace(line) &&
+        LrcStartTimeUtils.StartsWithLineTimeTag(line);
```
Comment (andy840119) [inline LrcParser/Parser/Lrc/Lines/LrcLyricParser.cs:15]: As I remembered, lyric text with no time-tag info can be treated as a lyric.
Feel free to find the evidence if I'm wrong.

File: LrcParser/Parser/Lrc/Utils/LrcStartTimeUtils.cs [hunk 2]
```diff
@@ -8,6 +8,19 @@ namespace LrcParser.Parser.Lrc.Utils;
...
 public static class LrcStartTimeUtils
 {
+    /// <summary>
+    /// Check if the line starts with a line time tag.
+    /// </summary>
+    /// <param name="line">The lyrics line in the LRC format.</param>
+    /// <returns>
+    /// <c>true</c> if the line starts with a line time tag, <c>false</c> otherwise.
+    /// </returns>
+    internal static bool StartsWithLineTimeTag(string line)
```
Comment (andy840119) [inline LrcParser/Parser/Lrc/Utils/LrcStartTimeUtils.cs:18]: If the previous comment is right, this is not needed also.

File: LrcParser.Tests/Parser/Lrc/LrcParserTest.cs [hunk 3]
```diff
@@ -31,7 +31,7 @@ public void TestDecode()
                         { new TextIndex(1), 18000 },
                         { new TextIndex(2), 19000 },
                         { new TextIndex(3), 20000 },
-                        { new TextIndex(3, IndexState.End), 21000 },
+                        { new TextIndex(4, IndexState.End), 21000 },
```
Comment (andy840119) [inline LrcParser.Tests/Parser/Lrc/LrcParserTest.cs:34]: Why this index changed?
Previous test case is incorrect?

File: LrcParser.Tests/Parser/Lrc/Lines/LrcLyricParserTest.cs [hunk 4]
```diff
@@ -39,16 +39,16 @@ public void TestDecode(string lyric, LrcLyric expected)
             {
                 Text = "帰り道は",
                 StartTimes = [17000],
-                TimeTags = TestCaseTagHelper.ParseTimeTags(["[0,start]:0", "[1,start]:1000", "[2,start]:2000", "[3,start]:3000", "[3,end]:4000"]),
+                TimeTags = TestCaseTagHelper.ParseTimeTags(["[0,start]:0", "[1,start]:1000", "[2,start]:2000", "[3,start]:3000", "[4,end]:4000"]),
             },
         ],
         [
-            "[00:17.00] 帰<00:01.00>り<00:02.00>道<00:03.00>は<00:04.00>",
+            "[00:01.00] 帰<00:01.10>り<00:02.00>道<00:03.00>は<00:04.00>",
```
Comment (andy840119) [inline LrcParser.Tests/Parser/Lrc/Lines/LrcLyricParserTest.cs:46]: What is the result if not change the test case?

File: LrcParser.Tests/Parser/Lrc/Lines/LrcLyricParserTest.cs [hunk 5]
```diff
@@ -39,16 +39,16 @@ public void TestDecode(string lyric, LrcLyric expected)
             {
                 Text = "帰り道は",
                 StartTimes = [17000],
-                TimeTags = TestCaseTagHelper.ParseTimeTags(["[0,start]:0", "[1,start]:1000", "[2,start]:2000", "[3,start]:3000", "[3,end]:4000"]),
+                TimeTags = TestCaseTagHelper.ParseTimeTags(["[0,start]:0", "[1,start]:1000", "[2,start]:2000", "[3,start]:3000", "[4,end]:4000"]),
```
Comment (andy840119) [inline LrcParser.Tests/Parser/Lrc/Lines/LrcLyricParserTest.cs:42]: Why did this number change?
Technically 3 is correct, TextIndex should be within 0(帰)~3(は)

**Review by andy840119 — COMMENTED** [review]
Only review the first and 3rd commit.
Resolve the review feedback than I'll continuous review the rest of one. Need to make sure that those breaking change is correct.

File: LrcParser/Parser/Lrc/Lines/LrcLyricParser.cs (hunk 1 above)
Comment (Maxr1998) [reply LrcParser/Parser/Lrc/Lines/LrcLyricParser.cs:15]: I think that in that case, the line doesn't follow LRC format, and shouldn't be parsed by the LRC parser. However, we can also change it so that the line time isn't strictly required.

The current `Decode` implementation already supports that and returns empty line start times and time tags with the raw text in that case. Just tell me what you prefer.

File: LrcParser.Tests/Parser/Lrc/Lines/LrcLyricParserTest.cs (hunk 5 above)
Comment (Maxr1998) [reply LrcParser.Tests/Parser/Lrc/Lines/LrcLyricParserTest.cs:42]: Same reasoning as above 👀

File: LrcParser/Parser/Lrc/Lines/LrcLyricParser.cs (hunk 1 above)
Comment (andy840119) [reply LrcParser/Parser/Lrc/Lines/LrcLyricParser.cs:15]: I prefer to remain the lyric for now.

Maybe in the next step, can:
- Write the validator to find the invalid part.
- Write the utils to fix or drop the lyric after being parsed to the model object (in LrcParser/Model)

File: LrcParser.Tests/Parser/Lrc/LrcParserTest.cs (hunk 3 above)
Comment (andy840119) [reply LrcParser.Tests/Parser/Lrc/LrcParserTest.cs:34]: The index in the text-index means the char index, not the gap index between chars.

image
Which means

When `{ new TextIndex(3, IndexState.End), 21000 }` means.
if the karaoke color mask goes to the end of the `は` character(full lyric is `帰り道は`), will at 21 second.

File: LrcParser.Tests/Parser/Lrc/Lines/LrcLyricParserTest.cs (hunk 4 above)
Comment (Maxr1998) [reply LrcParser.Tests/Parser/Lrc/Lines/LrcLyricParserTest.cs:46]: Exactly, that's why I changed the test case. The previous version was still wrong from the time where relative times were assumed.

File: LrcParser/Parser/Lrc/Lines/LrcLyricParser.cs (hunk 1 above)
Comment (Maxr1998) [reply LrcParser/Parser/Lrc/Lines/LrcLyricParser.cs:15]: Ok, sounds good.

File: LrcParser/Parser/Lrc/Utils/LrcStartTimeUtils.cs (hunk 2 above)
Comment (Maxr1998) [reply LrcParser/Parser/Lrc/Utils/LrcStartTimeUtils.cs:18]: Do we maybe want to keep this anyway and expose it to the API surface? Might be useful for users of the library. But we can also remove it again, your decision.

File: LrcParser/Parser/Lrc/Lines/LrcLyricParser.cs (hunk 1 above)
Comment (Maxr1998) [reply LrcParser/Parser/Lrc/Lines/LrcLyricParser.cs]: Fixed in da4c95b43a22af097c2cee2f61bac89919a2b5d0.

File: LrcParser/Parser/Lrc/Utils/LrcStartTimeUtils.cs (hunk 2 above)
Comment (andy840119) [reply LrcParser/Parser/Lrc/Utils/LrcStartTimeUtils.cs:18]: OK to keep this.

File: LrcParser.Tests/Parser/Lrc/Lines/LrcLyricParserTest.cs (hunk 4 above)
Comment (andy840119) [reply LrcParser.Tests/Parser/Lrc/Lines/LrcLyricParserTest.cs:46]: I might keep this test case as an edge case.

It's ok to make the refactor, but not change the test case.
Prefer to make it to another PR and describe why have those changes if have breaking change because lrc format has no strict spec.
Which means it's hard to decide whether to allow those changes merge or not if it's mixed with the refactor change.

**Review by andy840119 — COMMENTED** [review]
Seems there are several changes within a PR:
- Adjust test case.
- Doing refactor and add more tests.
- (breaking change) Deal with spacing within the lyric.
- (breaking change) Adjust the index.
- (breaking change but cancelled) not decode the lyric if not start with timing info.

As lrc has not strict rule definition, it's hard for me to decide lots of change within a PR.
Should be better to separate those changes with different PRs.

File: LrcParser.Tests/Parser/Lrc/Lines/LrcLyricParserTest.cs (hunk 5 above)
Comment (Maxr1998) [reply LrcParser.Tests/Parser/Lrc/Lines/LrcLyricParserTest.cs]: Reverted for now :+1:

File: LrcParser/Parser/Lrc/Lines/LrcLyricParser.cs [hunk 6]
```diff
@@ -9,13 +10,24 @@ namespace LrcParser.Parser.Lrc.Lines;
...
+        var (startTimes, rawLyric) = LrcStartTimeUtils.SplitLyricAndTimeTag(text);
+
+        // Word time tags can't be used if there are no or more than one line time tags
+        if (startTimes.Length is 0 or > 1)
+        {
+            return new LrcLyric
+            {
+                Text = rawLyric,
+                StartTimes = startTimes,
+                TimeTags = new SortedDictionary<TextIndex, int>()
```
[the remark this answers is not in the document]
Comment (Maxr1998) [reply LrcParser/Parser/Lrc/Lines/LrcLyricParser.cs:26]: The thing is, as those word time tags are always absolute, and not relative, it doesn't make sense to have multiple start time tags. They would just clash with each other. Thus, I don't think multiple start times can be supported if you have any word time tags.

[the remark this answers is not in the document]
Comment (Maxr1998) [reply LrcParser/Parser/Lrc/Lines/LrcLyricParser.cs:27]: Moved to #88.

File: LrcParser.Tests/Parser/Lrc/Utils/LrcTimedTextUtilsTest.cs [hunk 7]
```diff
@@ -12,26 +12,61 @@ public class LrcTimedTextUtilsTest
...
-
-        Assert.That(actualText, Is.EqualTo(expectedText));
-        Assert.That(actualTimeTags, Is.EqualTo(TestCaseTagHelper.ParseTimeTags(expectedTimeTags)));
-    }
-
-    [TestCase("<00:51.00><01:29.99><01:48.29><02:31.00><02:41.99>You gotta fight !", "You gotta fight !", new[] { "[0,start]:51000" })] // decode with invalid format.
-    public void TestDecodeWithInvalidFormat(string text, string expectedText, string[] expectedTimeTags)
-    {
-        var (actualText, actualTimeTags) = LrcTimedTextUtils.TimedTextToObject(text);
+        var (actualText, actualTimeTags) = LrcTimedTextUtils.TimedTextToObject(text, lineStartTime);
```
Comment (andy840119) [inline LrcParser.Tests/Parser/Lrc/Utils/LrcTimedTextUtilsTest.cs:69]: Have no idea why you remove this test case.

File: LrcParser.Tests/Parser/Lrc/Utils/LrcTimedTextUtilsTest.cs [hunk 8]
```diff
@@ -12,26 +12,61 @@ public class LrcTimedTextUtilsTest
...
+    [TestCase(
+        "Every <00:07.56>night <00:08.19>that <00:08.79>goes <00:09.59>between", 6840, "Every night that goes between",
+        new[] { "[0,start]:6840", "[6,start]:7560", "[12,start]:8190", "[17,start]:8790", "[22,start]:9590" }
+    )]
+    // Multiple empty tags
+    [TestCase(
+        "<00:51.00> <01:29.99><01:48.29>  <02:31.00> <02:41.99>You gotta fight !", 0,
+        "You gotta fight !", new[] { "[0,start]:161990" }
+    )]
+    public void TestDecode(string text, int lineStartTime, string expectedText, string[] expectedTimeTags)
```
Comment (andy840119) [inline LrcParser.Tests/Parser/Lrc/Utils/LrcTimedTextUtilsTest.cs:67]: Remove the lineStartTime because all the actual value should(and will) be zero.

File: LrcParser.Tests/Parser/Lrc/Lines/LrcLyricParserTest.cs [hunk 9]
```diff
@@ -50,7 +50,7 @@ public void TestDecode(string lyric, LrcLyric expected)
             {
                 Text = "帰り道は",
                 StartTimes = [17000],
-                TimeTags = TestCaseTagHelper.ParseTimeTags(["[1,start]:1000", "[2,start]:2000", "[3,start]:3000", "[3,end]:4000"]),
+                TimeTags = TestCaseTagHelper.ParseTimeTags(["[0,start]:17000", "[1,start]:1000", "[2,start]:2000", "[3,start]:3000", "[3,end]:4000"]),
```
Comment (andy840119) [inline LrcParser.Tests/Parser/Lrc/Lines/LrcLyricParserTest.cs:53]: nit:
add comment to mark that the first time-tag is auto-created.

**Review by andy840119 — APPROVED** [review]
Overall LGTM.

I haven't deeply review LrcParser/Parser/Lrc/Utils/LrcTimedTextUtils.cs but can give it a quick pass because the test case is much enough.

File: LrcParser.Tests/Parser/Lrc/Utils/LrcTimedTextUtilsTest.cs (hunk 8 above)
Comment (Maxr1998) [reply LrcParser.Tests/Parser/Lrc/Utils/LrcTimedTextUtilsTest.cs:67]: You mean just for this test? Should I split the tests where it's zero, maybe?

File: LrcParser.Tests/Parser/Lrc/Utils/LrcTimedTextUtilsTest.cs (hunk 7 above)
Comment (Maxr1998) [reply LrcParser.Tests/Parser/Lrc/Utils/LrcTimedTextUtilsTest.cs:69]: That's the one I moved above, since the parsing itself passes fine. Thus, I don't consider it invalid.

File: LrcParser.Tests/Parser/Lrc/Utils/LrcTimedTextUtilsTest.cs (hunk 8 above)
Comment (andy840119) [reply LrcParser.Tests/Parser/Lrc/Utils/LrcTimedTextUtilsTest.cs:67]: All test cases within this test lineStartTime should be 0.
You can make another test with few test case to test the result if lineStartTime is not zero if needed.

Comment (Maxr1998) [reply LrcParser.Tests/Parser/Lrc/Utils/LrcTimedTextUtilsTest.cs:67]: Ok, then I'll split it :+1:
Should also decrease the number of changed lines on the existing test.

File: LrcParser.Tests/Parser/Lrc/Utils/LrcTimedTextUtilsTest.cs (hunk 7 above)
Comment (Maxr1998) [reply LrcParser.Tests/Parser/Lrc/Utils/LrcTimedTextUtilsTest.cs:69]: Not really :smiling_face_with_tear:
However, the approach I chose seems logical to me:
`<00:51.00><01:29.99><01:48.29><02:31.00><02:41.99>You gotta fight !`
There are four segments that have a time defined, but no content. Thus, they don't need to be extracted and will be skipped. The last segment isn't empty, and will have a start time of `02:41.99`.

File: LrcParser.Tests/Parser/Lrc/Utils/LrcTimedTextUtilsTest.cs (hunk 8 above)
Comment (Maxr1998) [reply LrcParser.Tests/Parser/Lrc/Utils/LrcTimedTextUtilsTest.cs]: [Fixed](https://github.com/karaoke-dev/LrcParser/compare/c57f5e128cb7287b42443a7b0e611e0597dff2f8..81381f75679d8326ea6f7a88aba684c8cce8addb).
